text/0100-xcm-multi-type-asset-transfer.md

```diff
@@ -0,0 +1,282 @@
+| **Authors**     | Adrian Catangiu                                                                             |
+
+## Summary
+
+This RFC proposes a new instruction that provides a way to initiate on remote chains, asset transfers which
+transfer multiple types (teleports, local-reserve, destination-reserve) of assets, using XCM alone.
+
+The currently existing instructions are too opinionated and force each XCM asset transfer to a single
+transfer type (teleport, local-reserve, destination-reserve). This results in inability to combine different
+types of transfers in single transfer which results in overall poor UX when trying to move assets across
+chains.
+
+## Motivation
+
+XCM is the de-facto cross-chain messaging protocol within the Polkadot ecosystem, and cross-chain
+assets transfers is one of its main use-cases. Unfortunately, in its current spec, it does not support
+initiating on a remote chain, one or more transfers that combine assets with different transfer types.  
+For example, `ParachainA` cannot instruct `AssetHub` to teleport `ForeignAssetX` to `ParachainX` alongside
+`USDT` (which has to be reserve transferred) using current XCM specification.
+
+There currently exist `DepositReserveAsset`, `InitiateReserveWithdraw` and `InitiateTeleport` instructions
+that initiate asset transfers on execution, but they are opinionated in the type of transfer to use.
+Combining them is also not possible, because as a result of their individual execution, a message containing
+a `ClearOrigin` instruction is sent to the destination chain, making subsequent transfers impossible after
+the first instruction is executed.
+
+The new instruction proposed by this RFC allows an XCM program to describe multiple asset transfer types,
+then execute them in one shot with a single `remote_xcm` program sent to the target chain to effect
+the transfer and subsequently clear origin.
+
+Multi-hop asset transfers will benefit from this change by allowing single XCM program to handle multiple
+types of transfers and reduce complexity.
+
+Bridge asset transfers greatly benefit from this change by allowing building XCM programs to transfer multiple
+assets across multiple hops in a single pseudo-atomic action.  
+For example, allows single XCM program execution to transfer multiple assets from `ParaK` on Kusama, through Kusama
+Asset Hub, over the bridge through Polkadot Asset Hub with final destination `ParaP` on Polkadot.
+
+With current XCM, we are limited to doing multiple independent transfers for each individual hop in order to
+move both "interesting" assets, but also "supporting" assets (used to pay fees).
+
+**Note:** Transferring assets that require different paths (chains along the way) is _not supported within same XCM_ because of the async nature of cross chain messages. This new instruction, however, enables initiating transfers for multiple assets that take the same path even if they require different transfer types along that path.
```

Comment:
should we define the behaviour when not supported parameters are passed? I want to make sure there is no undefined behaviour.
e.g. if the implementation should check and reject it, or if this is implementation-specific behaviour (i.e. out of scope of RFC but should be documented by the XCM executor)

Reply:
The actual path messages take is out of scope for the XCVM, it only knows how to send them. The executor should document how it handles this

Reply:
I've added more details and clearly specified that this instruction is used point-to-point, and has to be composed to achieve multi-hop transfers.

For each hop, its parameters unequivocally specify what should happen for that specific hop/leg, so there should be no room for undefined behavior.

Any chain-specific checks or validations of the parameters are implementation-specific behavior.